#### ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
### AUTHOR - CHANDAN D. CHAUDHARI

#### DATASETLINK - https://raw.githubusercontent.com/chandanc5525/Dataset/main/Data/telecom_churn.csv 
#### ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------

In [1]:
!pip install klib

Defaulting to user installation because normal site-packages is not writeable


### PROBLEM STATEMENT

Customer churn is a major problem and one of the most important concerns for large companies. Due to the direct effect on the revenues of the companies, especially in the telecom feld, companies are seeking to develop means to predict potential customer to churn. Therefore, fnding factors that increase customer churn is important to take necessary actions to reduce this churn.

In [2]:
# Importing Python Libraries
import numpy as np
import pandas as pd

# Importing Data Visualization Libraries
import seaborn as sns
import matplotlib.pyplot as plt

# Import Filter Warning Libraries
import warnings
warnings.filterwarnings('ignore')

In [3]:
# Importing Dataset

URL = "https://raw.githubusercontent.com/chandanc5525/Dataset/main/Data/"

df = pd.read_csv(URL + 'telecom_churn.csv')

# Checking random 10 Sample of the dataset
df.sample(10)

State  Account length  Area code International plan Voice mail plan  \
2138    WY             124        415                 No              No   
1331    OK             101        408                 No              No   
1493    MT             126        415                 No             Yes   
448     GA              75        415                 No              No   
2746    CO             107        415                 No              No   
1538    WV             113        415                 No              No   
76      DC              82        415                 No              No   
2336    NY             122        415                 No             Yes   
1438    ND              60        510                 No              No   
1870    AZ              43        415                 No              No   

      Number vmail messages  Total day minutes  Total day calls  \
2138                      0              178.3              102   
1331                      0              209.6              107   
1493                     30              153.4               90   
448                       0              117.3              114   
2746                      0              204.5              108   
1538                      0               61.2              111   
76                        0              300.3              109   
2336                     37              163.0              107   
1438                      0              203.2               99   
1870                      0              179.3               97   

      Total day charge  Total eve minutes  Total eve calls  Total eve charge  \
2138             30.31              235.0              120             19.98   
1331             35.63              228.8               96             19.45   
1493             26.08              151.4               97             12.87   
448              19.94              201.1               61             17.09   
2746             34.77              162.4              110             13.80   
1538             10.40               92.3               88              7.85   
76               51.05              181.0              100             15.39   
2336             27.71              312.8              118             26.59   
1438             34.54              235.8              131             20.04   
1870             30.48              252.7              126             21.48   

      Total night minutes  Total night calls  Total night charge  \
2138                239.7                119               10.79   
1331                172.4                 85                7.76   
1493                153.8                 97                6.92   
448                 107.9                 82                4.86   
2746                155.0                102                6.98   
1538                197.4                114                8.88   
76                  270.1                 73               12.15   
2336                200.0                 85                9.00   
1438                224.9                112               10.12   
1870                227.5                114               10.24   

      Total intl minutes  Total intl calls  Total intl charge  \
2138                10.9                 1               2.94   
1331                 7.6                 2               2.05   
1493                12.8                 4               3.46   
448                 12.2                 3               3.29   
2746                13.4                 1               3.62   
1538                13.7                 3               3.70   
76                  11.7                 4               3.16   
2336                11.6                 5               3.13   
1438                15.1                 6               4.08   
1870                 8.0                 5               2.16   

      Customer service calls  Churn  
2138                       3  False  
1331    

In [4]:
# Checking Shape of the dataset
df.shape

(3333, 20)

In [5]:
# Checking Missing values in the dataset

df.isnull().sum()

State                     0
Account length            0
Area code                 0
International plan        0
Voice mail plan           0
Number vmail messages     0
Total day minutes         0
Total day calls           0
Total day charge          0
Total eve minutes         0
Total eve calls           0
Total eve charge          0
Total night minutes       0
Total night calls         0
Total night charge        0
Total intl minutes        0
Total intl calls          0
Total intl charge         0
Customer service calls    0
Churn                     0
dtype: int64

In [6]:
# Checking Descriptive Stats for Categorical Column

df.describe(include='all')

State  Account length    Area code International plan Voice mail plan  \
count   3333     3333.000000  3333.000000               3333            3333   
unique    51             NaN          NaN                  2               2   
top       WV             NaN          NaN                 No              No   
freq     106             NaN          NaN               3010            2411   
mean     NaN      101.064806   437.182418                NaN             NaN   
std      NaN       39.822106    42.371290                NaN             NaN   
min      NaN        1.000000   408.000000                NaN             NaN   
25%      NaN       74.000000   408.000000                NaN             NaN   
50%      NaN      101.000000   415.000000                NaN             NaN   
75%      NaN      127.000000   510.000000                NaN             NaN   
max      NaN      243.000000   510.000000                NaN             NaN   

        Number vmail messages  Total day minutes  Total day calls  \
count             3333.000000        3333.000000      3333.000000   
unique                    NaN                NaN              NaN   
top                       NaN                NaN              NaN   
freq                      NaN                NaN              NaN   
mean                 8.099010         179.775098       100.435644   
std                 13.688365          54.467389        20.069084   
min                  0.000000           0.000000         0.000000   
25%                  0.000000         143.700000        87.000000   
50%                  0.000000         179.400000       101.000000   
75%                 20.000000         216.400000       114.000000   
max                 51.000000         350.800000       165.000000   

        Total day charge  Total eve minutes  Total eve calls  \
count        3333.000000        3333.000000      3333.000000   
unique               NaN                NaN              NaN   
top                  NaN                NaN              NaN   
freq                 NaN                NaN              NaN   
mean           30.562307         200.980348       100.114311   
std             9.259435          50.713844        19.922625   
min             0.000000           0.000000         0.000000   
25%            24.430000         166.600000        87.000000   
50%            30.500000         201.400000       100.000000   
75%            36.790000         235.300000       114.000000   
max            59.640000         363.700000       170.000000   

        Total eve charge  Total night minutes  Total night calls  \
count        3333.000000          3333.000000        3333.000000   
unique               NaN                  NaN                NaN   
top                  NaN                  NaN                NaN   
freq                 NaN                  NaN                NaN   
mean           17.083540           200.872037         100.107711   
std             4.310668            50.573847          19.568609   
min             0.000000            23.200000          33.000000   
25%            14.160000           167.000000          87.000000   
50%            17.120000           201.200000         100.000000   
75%            20.000000           235.300000         113.000000   
max            30.910000           395.000000         175.000000   

        Total night charge  Total intl minutes  Total intl calls  \
count          3333.000000         3333.000000       3333.000000   
unique                 NaN                 NaN               NaN   
top                    NaN                 NaN               NaN   
freq                   NaN                 NaN               NaN   
mean              9.039325           10.237294          4.479448   
std               2.275873            2.791840          2.461214   
min               1.040000            0.000000          0.000000   
25%               7.520000            8.500000          3.000000   
50%               9

In [7]:
# Feature Columns

df.columns

Index(['State', 'Account length', 'Area code', 'International plan',
       'Voice mail plan', 'Number vmail messages', 'Total day minutes',
       'Total day calls', 'Total day charge', 'Total eve minutes',
       'Total eve calls', 'Total eve charge', 'Total night minutes',
       'Total night calls', 'Total night charge', 'Total intl minutes',
       'Total intl calls', 'Total intl charge', 'Customer service calls',
       'Churn'],
      dtype='object')

In [8]:
# Checking Information about the Dataset
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3333 entries, 0 to 3332
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   State                   3333 non-null   object 
 1   Account length          3333 non-null   int64  
 2   Area code               3333 non-null   int64  
 3   International plan      3333 non-null   object 
 4   Voice mail plan         3333 non-null   object 
 5   Number vmail messages   3333 non-null   int64  
 6   Total day minutes       3333 non-null   float64
 7   Total day calls         3333 non-null   int64  
 8   Total day charge        3333 non-null   float64
 9   Total eve minutes       3333 non-null   float64
 10  Total eve calls         3333 non-null   int64  
 11  Total eve charge        3333 non-null   float64
 12  Total night minutes     3333 non-null   float64
 13  Total night calls       3333 non-null   int64  
 14  Total night charge      3333 non-null   

In [9]:
features = df[['State','International plan','Voice mail plan','Churn']]
for i in features:
    print('-----------------------')
    print(i)
    print(features[i].unique())
    print(features[i].value_counts())
    print('-----------------------')

-----------------------
State
['KS' 'OH' 'NJ' 'OK' 'AL' 'MA' 'MO' 'LA' 'WV' 'IN' 'RI' 'IA' 'MT' 'NY'
 'ID' 'VT' 'VA' 'TX' 'FL' 'CO' 'AZ' 'SC' 'NE' 'WY' 'HI' 'IL' 'NH' 'GA'
 'AK' 'MD' 'AR' 'WI' 'OR' 'MI' 'DE' 'UT' 'CA' 'MN' 'SD' 'NC' 'WA' 'NM'
 'NV' 'DC' 'KY' 'ME' 'MS' 'TN' 'PA' 'CT' 'ND']
WV    106
MN     84
NY     83
AL     80
WI     78
OH     78
OR     78
WY     77
VA     77
CT     74
MI     73
ID     73
VT     73
TX     72
UT     72
IN     71
MD     70
KS     70
NC     68
NJ     68
MT     68
CO     66
NV     66
WA     66
RI     65
MA     65
MS     65
AZ     64
FL     63
MO     63
NM     62
ME     62
ND     62
NE     61
OK     61
DE     61
SC     60
SD     60
KY     59
IL     58
NH     56
AR     55
GA     54
DC     54
HI     53
TN     53
AK     52
LA     51
PA     45
IA     44
CA     34
Name: State, dtype: int64
-----------------------
-----------------------
International plan
['No' 'Yes']
No     3010
Yes     323
Name: International plan, dtype: int64
-----------------------
-------

### OBSERVATIONS

1. THE ABOVE DATASET CONTAINS 3333 ROWS AND 20 COLUMNS
2. NULL VALUES ARE ABSENT IN THE GIVEN DATASET
3. CHURN COLUMN IS BOOLEANS, SO WE NEED TO CONVERT INTO NUMERICAL FORM